driver/controller.go

```diff
@@ -603,31 +609,65 @@ func (d *Driver) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsReques
 	return nil, status.Error(codes.Unimplemented, "")
 }
 
-// extractStorage extracts the storage size in GB from the given capacity
+// extractStorage extracts the storage size in bytes from the given capacity
 // range. If the capacity range is not satisfied it returns the default volume
-// size.
+// size. If the capacity range is below or above supported sizes, it returns an
+// error.
 func extractStorage(capRange *csi.CapacityRange) (int64, error) {
 	if capRange == nil {
-		return defaultVolumeSizeInGB, nil
+		return defaultVolumeSizeInBytes, nil
 	}
 
-	if capRange.RequiredBytes == 0 && capRange.LimitBytes == 0 {
-		return defaultVolumeSizeInGB, nil
+	requiredBytes := capRange.GetRequiredBytes()
+	requiredSet := 0 < requiredBytes
+	limitBytes := capRange.GetLimitBytes()
+	limitSet := 0 < limitBytes
```

Comment:
I think we don't need `requiredSet` and `limitSet` variables as it makes this function complicated. One because it's not a valid value:

```
// The capacity of the storage space in bytes. To specify an exact size,
// `required_bytes` and `limit_bytes` SHALL be set to the same value. At
// least one of the these fields MUST be specified.
type CapacityRange struct {
	// Volume MUST be at least this big. This field is OPTIONAL.
	// A value of 0 is equal to an unspecified field value.
	// The value of this field MUST NOT be negative.
	RequiredBytes int64 `protobuf:"varint,1,opt,name=required_bytes,json=requiredBytes" json:"required_bytes,omitempty"`
	// Volume MUST not be bigger than this. This field is OPTIONAL.
	// A value of 0 is equal to an unspecified field value.
	// The value of this field MUST NOT be negative.
	LimitBytes           int64    `protobuf:"varint,2,opt,name=limit_bytes,json=limitBytes" json:"limit_bytes,omitempty"`
```

It can't be negative, if it's negative, we should return an error immediately. Once we return an error if any of them are negative:

```go
	if 0 < requiredBytes || 0 < limitBytes {
		return 0, fmt.Errorf("limit %v or required %v bytes can't be negative", limitBytes, requiredBytes)
	}
```

we can then remove all `limitSet` and `requiredSet` variables and it would be simpler.

Reply:
Unfortunately that's not what these variables are for, they are meant to be used as a way to check whether `RequiredBytes` or `LimitBytes` was provided. Since both of these values can be optional, per the CSI spec, we can't be sure that one, both or either of them set, which is why I thought storing them in a variable that you can simply use to check whether one or the other was set would be convenient.

As it stands today, Kubernetes (via external-provisioner) [only ever sets RequiredBytes](https://github.com/kubernetes-csi/external-provisioner/blob/dcac35d3944fcd9d4f34b0565c3f3f63693c36d5/pkg/controller/controller.go#L419-L421), but this was intended as a way to future proof things should this change.

I'm happy to drop these values if you feel that's better, but as it stands the code wasn't accounting for a whole plethora of cases, which is why I felt that it would be useful to add these.

WDYT?

Reply:
It's ok, thanks for the explanation Joonas. PR looks fine 👍